### DISCUSSION TITLE
Skinning native USD meshes

### DISCUSSION THREAD

**Discussion by nickyliux** [Autodesk/maya-usd#2073] [Q&A]
1) I'm a character rigger and learning about USD for maya only very lately. As a test I referenced in character model in USD format and rigged it as usual. But I feel that I'm missing great features of USD if not loaded natively. And natively loaded object can't be skinned. Is it the limitation of current version of USD for maya or am I missing something ?

2) I see things like USDSkel and skinning but found no info of application in maya. Just curious of their stage of development and usage.

**Comment by kxl-adsk** [comment]
Great questions @nickyliux. You are absolutely right that there is no general solution to rigging, nor to execution/computation in USD. It's mentioned on the intro page
> [Not an execution or rigging system](https://graphics.pixar.com/usd/release/intro.html#not-an-execution-or-rigging-system)
USD provides a lightweight, optimized scenegraph to facilitate authoring and efficient extraction of composed scene description. However, it provides no other behaviors than composition of a namespace hierarchy and property [Value Resolution](https://graphics.pixar.com/usd/release/glossary.html#usdglossary-valueresolution), and in the tradeoff space between “low-memory footprint, higher-latency data access” and “high-memory footprint, low-latency access to data”, USD’s scenegraph leans more towards the former, whereas a high-performance execution engine requires the latter.
>
> Further, the more rigging behaviors and execution semantics we would add to USD, the more difficult it would become to interchange the data successfully between DCC’s, since there is not, currently, broad agreement between vendors of what these behaviors should be.
>
> USD and its schema generation tools should be suitable for encoding rigging for round-tripping rigging data in a particular application or custom pipeline, and USD does provide facilities that a client could use to build more extensive in-memory caches on top of a UsdStage to provide lower-latency access to data encoded in USD. But for now, these do not play a significant role in what we feel is the primary directive of USD: scalable interchange of geometric, shading, and lighting data between DCC’s in a 3D content creation pipeline.

We have in the works an approach that lets you leverage Maya's data model for editing and animation. It's highly inspired by the MayaReferene workflow in the AnimalLogic plugin, where you can switch between Maya's rig and cached representation.

Regarding UsdSkel - I invite others to comment on usage, TMK UsdSkel is mostly used for crowds. We have it on our radar and recently added VP2RenderDelegate support for it. Some other pieces have been contributed by the community (like blendshape export contributed by Apple).
